cmd/dagger/functions.go

```diff
@@ -0,0 +1,546 @@
+			cmd.SetContext(ctx)
+			rec := progrock.FromContext(ctx)
+
+			c.mod.LoadObject(fn.ReturnType)
+
+			for _, arg := range fn.Args {
+				c.mod.LoadObject(arg.TypeDef)
+			}
+
+			argValues := make(map[string]any)
+
+			for _, arg := range fn.Args {
+				p, err := arg.AddFlag(cmd.Flags(), dag)
+				if err != nil {
+					return fmt.Errorf("set flag %q: %w", arg.FlagName(), err)
+				}
+				argValues[arg.Name] = p
+
+				// TODO: This won't work on its own because cobra checks required
+				// flags before the RunE function is called. We need to manually
+				// check the flags after parsing them.
+				if !arg.TypeDef.Optional {
+					cmd.MarkFlagRequired(arg.FlagName())
+				}
+			}
+
+			if fc.OnSetFlags != nil {
+				return fc.OnSetFlags(cmd, fn.ReturnType)
+			}
+
+			if err := cmd.ParseFlags(args); err != nil {
+				return fmt.Errorf("parse sub-command flags: %w", err)
+			}
+
+			fc.addSubCommands(ctx, dag, fn.ReturnType.AsObject, c, cmd)
+
+			// Need to make the query selection before chaining off.
+			returnType, err := fc.selectFunc(fn, c, argValues, dag)
+			if err != nil {
+				return fmt.Errorf("query selection: %w", err)
+			}
+
+			help, _ := cmd.Flags().GetBool("help")
+			cmdArgs := cmd.Flags().Args()
+
+			subCmd, subArgs, err := cmd.Find(cmdArgs)
+			if err != nil {
+				return fmt.Errorf("find sub-command: %w", err)
+			}
+
+			// If a subcommand was found, run it instead.
+			if subCmd != cmd {
+				return fc.run(subCmd, subArgs)
+			}
+
+			if fc.CheckReturnType != nil {
+				if err = fc.CheckReturnType(c, returnType); err != nil {
+					return err
+				}
+			}
+
+			c.loader.Complete()
+
+			vtx := rec.Vertex(
+				"cmd-func-exec",
+				fmt.Sprintf("running %q", cmd.CommandPath()),
+				progrock.Focused(),
+			)
+			defer func() { vtx.Done(err) }()
+
+			setCmdOutput(cmd, vtx)
+
+			// Make sure --help is running on the leaf command (subCmd == cmd),
+			// otherwise the first command in the chain swallows it.
+			if help {
+				return cmd.Help()
+			}
+
+			query, _ := c.q.Build(ctx)
+			rec.Debug("executing", progrock.Labelf("query", "%+v", query))
+
+			var response any
+
+			q := c.q.Bind(&response)
+
+			if err := q.Execute(ctx, c.c); err != nil {
+				return err
+			}
+
+			if fc.OnResult != nil {
+				return fc.OnResult(c, cmd, *returnType, &response)
+			}
+
+			if returnType.Kind != dagger.Voidkind {
+				cmd.Println(response)
+			}
+
+			return nil
+		},
+	}
+
+	if fn.Name != newCmd.Name() {
+		newCmd.Aliases = append(newCmd.Aliases, fn.Name)
+	}
+
+	newCmd.InitDefaultHelpFlag()
+	newCmd.Flags().SetInterspersed(false)
+
+	return newCmd
+}
+
+// selectFunc adds the function selection to the query.
+// Note that the type can change if there's an extra selection for supported types.
+func (fc *FuncCommand) selectFunc(fn *modFunction, c *callContext, argValues map[string]any, dag *dagger.Client) (*modTypeDef, error) {
+	c.Select(fn.Name)
+
+	// TODO: Check required flags.
+	// TODO: Handle default values.
```

Comment:
I'm actually not sure we do need to handle default values here ever; I *think* they get handled by our graphql server (and if they don't we should still handle setting them server-side)

Reply:
That's a good point. I just need to check if the flag has been changed or not because right now the flags are set with an empty value and if not set that's what will be sent here.

Reply:
This will be handled more gracefully in a follow up as it's already a part of an ongoing refactor.